File: 프로그래머스/unrated/155651. 호텔 대실/호텔 대실.py
from heapq import heappush, heappop
def solution(book_time):
    """19:12"""
    answer = 0
    newTime = []
    for start, end in book_time:
        start_ = start.split(":")
        end_ = end.split(":")
        newTime.append([int(start_[0])*60 + int(start_[1]), int(end_[0])*60+int(end_[1])])
    newTime.sort(key=lambda x : x[0] * 1440 + x[1])
    rooms = []
    for start, end in newTime:
        # print(rooms)
        while rooms:
            if rooms[0] <= start:
                heappop(rooms)
            else:
                break
        heappush(rooms, end+10)
        if answer < len(rooms):
            answer = len(rooms)
        
    return answer

File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.java
import java.util.*;

class Solution {
    int answer;
    int n;
    Reservation[] reservations;
    PriorityQueue<Reservation> heap;

    
    public int solution(String[][] book_time) {
        init(book_time);
        solve();
        return answer;
    }
    
    void init(String[][] bookTime) {
        n = bookTime.length;
        reservations = new Reservation[n];
        heap = new PriorityQueue<>(Reservation::sortWithEnd);
        for (int i = 0; i < n; i++) {
            String[] book = bookTime[i];
            reservations[i] = new Reservation(strToInt(book[0]), strToInt(book[1])+10);
        }
        Arrays.sort(reservations, Reservation::sortWithStart);
    }
    
    void solve() {
        for (Reservation reservation : reservations) {
            while (!heap.isEmpty() && reservation.start >= heap.peek().end) {
                heap.remove();
            }
            heap.add(reservation);
            answer = Math.max(answer, heap.size());
        }
    }
    
    int strToInt(String str) {
        String[] strArray = str.split(":");
        return 60*Integer.parseInt(strArray[0]) + Integer.parseInt(strArray[1]);
    }
    
    static class Reservation {
        int start;
        int end;
        
        Reservation(int start, int end) {
            this.start = start;
            this.end = end;
        }
        
        int sortWithStart(Reservation compare) {
            return Integer.compare(this.start, compare.start);
        }
        
        int sortWithEnd(Reservation compare) {
            return Integer.compare(this.end, compare.end);
        }
    }
}